### PR TITLE
[KeyVault] updating KeyVaultAuthBase to reuse original request connection

### PR DESCRIPTION
**Pull request by schaabs**

### CODE REVIEW

**Comment by schaabs** [comment]
This update is intended to fix issue https://github.com/Azure/azure-sdk-for-python/issues/1159

**Comment by RandalliLama** [comment]
@schaabs Is the authentication callback signature doc'ed anywhere?

File: azure-keyvault/azure/keyvault/custom/key_vault_authentication.py [hunk 1]
```diff
@@ -19,27 +16,92 @@ class KeyVaultAuthBase(AuthBase):
...
+            self._thread_local.auth_attempted = False
+
+    def handle_401(self, response, **kwargs):
+        """
+        Takes the response authenticates and resends if neccissary
+        
+        :return: The final response to the authenticated request
+        :rtype: requests.Response
+        """
+        # If response is not 4xx do not auth and return response
```
Comment (dragav) [inline azure-keyvault/azure/keyvault/custom/key_vault_authentication.py]: Shouldn't the response be exactly 401? I.e. a 403 is not a valid code when responding with a challenge.

File: azure-keyvault/azure/keyvault/custom/key_vault_authentication.py [hunk 2]
```diff
@@ -19,27 +16,92 @@ class KeyVaultAuthBase(AuthBase):
...
+            return response
+
+        # If we've already attempted to auth for this request once, do not auth and return response
+        if self._thread_local.auth_attempted:
+            self._thread_local.auth_attempted = False
+            return response
+
+        # Otherwise authenticate and retry the request
+        self._thread_local.auth_attempted = True
+
```
Comment (dragav) [inline azure-keyvault/azure/keyvault/custom/key_vault_authentication.py]: Would it make sense to move this toggle (that authn was attempted) after the successful submission of the response?

File: azure-keyvault/azure/keyvault/custom/key_vault_authentication.py (hunk 1 above)
Comment (schaabs) [inline azure-keyvault/azure/keyvault/custom/key_vault_authentication.py]: In [RFC 7235](https://tools.ietf.org/html/rfc7235#section-4.1)  it says:  A server MAY generate a WWW-Authenticate header field in other response messages to indicate that supplying credentials (or different credentials) might affect the response.  So I was going by that and what DigestAuth is doing in the request library https://github.com/requests/requests/blob/master/requests/auth.py#L231.  But if the data plane service will only challenge with a 401 response code we can restrict this to 401.

File: azure-keyvault/azure/keyvault/custom/key_vault_authentication.py (hunk 2 above)
Comment (schaabs) [inline azure-keyvault/azure/keyvault/custom/key_vault_authentication.py]: It has to be done before we resubmit the request to handle the case that the resubmitted request redirects.

**Comment by RandalliLama** [comment]
@schaabs Doc strings was all I had in mind.

**Pull request merged**